pubsub/alicloud/confluentkafka/confluentkafka.go

```diff
@@ -0,0 +1,266 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation and Dapr Contributors.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package confluentkafka
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strconv"
+
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"github.com/dapr/components-contrib/pubsub"
+	"github.com/dapr/kit/logger"
+)
+
+var (
+	ProducerFlushTimeoutMs = 10 * 1000
+)
+
+type contextWithCancelFunc struct {
+	Ctx        context.Context
+	CancelFunc context.CancelFunc
+}
+
+type Kafka struct {
+	ConfigMap *kafka.ConfigMap
+	logger    logger.Logger
+	producer  *kafka.Producer
+	consumer  *kafka.Consumer
+	topics    []string
+
+	consumerContext contextWithCancelFunc
+}
+
+func NewConfluentKafka(logger logger.Logger) pubsub.PubSub {
+	return &Kafka{ //nolint:exhaustivestruct
+		logger: logger,
+	}
+}
+
+func (k *Kafka) parseMetadata(props map[string]string) error {
+	k.ConfigMap = &kafka.ConfigMap{}
+
+	var metadata map[string]interface{}
+	if configJson, ok := props["configJson"]; ok {
+		err := json.Unmarshal([]byte(configJson), &metadata)
+		if err != nil {
+			return err
+		}
+
+		for key, val := range metadata{
+			if reflect.ValueOf(val).Kind() == reflect.Float64{
+				err = k.ConfigMap.SetKey(key, int(val.(float64)))
+				if err != nil {
+					return err
+				}
+			} else {
+				err = k.ConfigMap.SetKey(key, val)
+				if err != nil {
+					return err
+				}
+			}
+		}
+
+		return nil
+	} else {
+		return fmt.Errorf("needs key configJson in metadata for initialization and check https://github.com/confluentinc/confluent-kafka-go for format of the config")
+	}
+}
+
+func (k *Kafka) Init(md pubsub.Metadata) error {
+	err := k.parseMetadata(md.Properties)
+	if err != nil {
+		return err
+	}
+
+	k.producer, err = kafka.NewProducer(k.ConfigMap)
+	if err != nil {
+		return err
+	}
+
+	k.consumer, err = kafka.NewConsumer(k.ConfigMap)
```

Comment:
Should initializing the consumer be moved to `Subscribe`? (Perhaps surrounded by a `sync.Once`)?

Reply:
When sync.Once is used for initializing consumer in Subscribe, the error might be missed if Subscribe is called multiple times, and it will be very confusing.